packages/unified-selection/src/unified-selection/HiliteSetProvider.ts

```diff
@@ -0,0 +1,338 @@
+
+    array.push(key.id);
+    map.set(keyType, array);
+  }
+
+  private async getType(key: SelectableInstanceKey): Promise<InstanceIdType> {
+    const cachedType = this._classRelationCache.get(key.className);
+    if (cachedType) {
+      return cachedType;
+    }
+
+    const keyClass = await this.getClass(key.className);
+    if (!keyClass) {
+      return "unknown";
+    }
+
+    if (this._hasFunctionalSchema) {
+      const functionalType = await this.checkType(key, keyClass, "Functional", "FunctionalElement", "functionalElement");
+      if (functionalType) {
+        return functionalType;
+      }
+    }
+
+    return (
+      (await this.checkType(key, keyClass, "BisCore", "Subject", "subject")) ??
+      (await this.checkType(key, keyClass, "BisCore", "Model", "model")) ??
+      (await this.checkType(key, keyClass, "BisCore", "Category", "category")) ??
+      (await this.checkType(key, keyClass, "BisCore", "SubCategory", "subCategory")) ??
+      (await this.checkType(key, keyClass, "BisCore", "GroupInformationElement", "groupInformationElement")) ??
+      (await this.checkType(key, keyClass, "BisCore", "GeometricElement", "geometricElement")) ??
+      (await this.checkType(key, keyClass, "BisCore", "Element", "element")) ??
+      "unknown"
+    );
+  }
+
+  private async checkType(key: SelectableInstanceKey, keyClass: ECClass, schemaName: string, className: string, type: InstanceIdType) {
+    if (await keyClass.is(className, schemaName)) {
+      this._classRelationCache.set(key.className, type);
+      return type;
+    }
+    return undefined;
+  }
+
+  private async getHilitedModels(map: Map<InstanceIdType, string[]>): Promise<string[]> {
+    const modelKeys = map.get("model") ?? [];
+    const subjectKeys = map.get("subject") ?? [];
+
+    if (modelKeys.length === 0 && subjectKeys.length === 0) {
+      return [];
+    }
+
+    const bindings: ECSqlBinding[] = [
+      { type: "idset", value: subjectKeys },
+      { type: "idset", value: modelKeys },
+    ];
+    const query = `WITH RECURSIVE
+                    ChildSubjects (ECInstanceId) AS (
+                      SELECT ECInstanceId FROM BisCore.Subject WHERE InVirtualSet(?, ECInstanceId)
+                      UNION ALL
+                      SELECT r.TargetECInstanceId FROM ChildSubjects s
+                        JOIN BisCore.ElementOwnsChildElements r ON r.SourceECInstanceId = s.ECInstanceId
+                      WHERE r.TargetECClassId IS (BisCore.Subject)
+                    ),
+                    Models (ECInstanceId) AS (
+                      SELECT s.ECInstanceId as \`ECInstanceId\` FROM BisCore.Model s
+                      WHERE InVirtualSet(?, ECInstanceId)
+                    )
+                    SELECT r.TargetECInstanceId as \`ECInstanceId\` FROM ChildSubjects s
+                      JOIN BisCore.ElementOwnsChildElements r ON r.SourceECInstanceId = s.ECInstanceId
+                      WHERE r.TargetECClassId IS (BisCore.PhysicalPartition)
```

Comment:
would be more efficient to join `bis.PhysicalPartition` using its `Parent` navigation property